### ISSUE TITLE
PaxExam's issues with mvn protocol handler regarding download of karaf distribution [PAXEXAM-799]

### ISSUE THREAD

**Issue by ops4j-issues** [ops4j/org.ops4j.pax.exam2#892]
**[Piotr Skowronek](https://ops4j1.jira.com/secure/ViewProfile.jspa?accountId=557058:1971503f-8ca4-4b91-9246-1a5f33102d3d)** created **[PAXEXAM-799](https://ops4j1.jira.com/browse/PAXEXAM-799)** 

I discovered several issues with PaxExam and its maven handling regarding download of Karaf Distribution (it may also affect other containers).<br /> This is going be quite long report, but I didn't want to split the issue so one could see the whole picture.

1. PaxExam uses java.net.URL with mvn protocol handler (org.ops4j.pax.url.mvn.Handler) to download maven artifacts - apparently this handler does not handle redirections properly. I found out about that when our tests started to fail out of the blue - there was no karaf distribution extracted. The cause was hard to investigate since there was no relevant logs and I had to debug through the code to see that the handler contained hardcoded list of artifact repositories and the first one on the list was: [http://osgi.sonatype.org](http://osgi.sonatype.org) repository - which apparently started to return 301 when trying to download karaf distribution zip.

2. In such a case URL.openStream simply returns html with 301 which is then given to org.apache.commons.compress.archivers.ArchiveInputStream which tries to unzip the stream to a directory where Karaf base should appear. Since obviously the stream is not ZIP it silently does nothing. The result is that PaxExam fails as the Karaf base does not exist. The logs only contain:

```
java.lang.IllegalStateException: No karaf base dir found in extracted distribution.
```

and nothing else which could shed some light as what was the reason.

3. To overcome the situation I had to reconfigure the handler's repository URLs to point at our Artifactory repository by using the following system properties:

```
org.ops4j.pax.url.mvn.defaultRepositories
org.ops4j.pax.url.mvn.repositories
```

This is another issue that one must use system properties to do that and it cannot be done via PaxExam configuration (i.e. CoreOptions.repositories() ) - org.ops4j.pax.url.mvn.Handler does not respect that.

4. Having the repositories overridden an another problem appeared - the protocol handler uses org.ops4j.net.URLUtils which has hardcoded user agent of ... Safari browser (details [here](https://ops4j1.jira.com/browse/BASE-48)). This cannot be overridden and it causes Artifactory's maven repo to redirect any request to log in page.... No matter if artifact exists or not - Artifactory sees a browser-like request and redirects to log in page. Of course if the user agent is not browser-like or the header is not set at all it returns the requested artifact.<br /> Since the redirection is not handled (point 1) and nothing is logged (point 2) one more time we don't know why PaxExam fails to prepare Karaf base.

To recap, I see 5 problems:<br /> 1. maven handler to does not handle redirections<br /> 2. the code which handles the stream does not check if it is of expected content type (kinda hard when using URL to get the stream...)<br /> 3. no logs from ArchiveInputStream<br /> 4. maven protocol handler uses its own list of repositories and it does not respect PaxExam configuration (or rather PaxExam configuration does not configure mvn handler)<br /> 5. URLUtils having hard coded user agent of Safari:

* the user agent should be non-Safari like - it should reflect the library somehow
* the user agent should be configurable, maybe via System property "user-agent"?

---

**Affects:** 4.5.0, 4.9.1
Votes: 0, Watches: 1
